[libc++] LWG-4021 "`mdspan::is_always_meow()` should be `noexcept`", use `LIBCPP_STATIC_ASSERT` for `noexcept` strengthening

StephanTLavavej opened the pull request:
Found while running libc++'s test suite with MSVC's STL.

* I've filed [LWG-4021](https://cplusplus.github.io/LWG/issue4021) "`mdspan::is_always_meow()` should be `noexcept`" and implemented this in libc++'s product and test code.
* Use `LIBCPP_STATIC_ASSERT` to avoid issues with `noexcept` strengthening in MSVC's STL.
  + As permitted by the Standard, MSVC's STL conditionally strengthens `mdspan` construction/`is_meow`/`stride` and `elements_view` iterator `base() &&`, and always strengthens `basic_stringbuf` `swap`.
  + In `mdspan/properties.pass.cpp`, this also upgrades runtime `assert`s to `static_assert`s.
* Improvement: Upgrade `assert` to `static_assert` when inspecting the `noexcept`ness of `std::ranges::iter_move`. (These `!noexcept` tests weren't causing issues for MSVC's STL, so I didn't change them to be libc++-specific.)

StephanTLavavej committed:
Use `LIBCPP_STATIC_ASSERT` to avoid issues with noexcept strengthening in MSVC's STL.

MSVC's STL conditionally strengthens `mdspan` construction/`is_meow`/`stride` and `elements_view` iterator `base() &&`,
and always strengthens `mdspan` `is_always_meow` and `basic_stringbuf` `swap`, as permitted by the Standard.

StephanTLavavej committed:
Improvement: Upgrade assert to static_assert when inspecting noexcept.

philnik777 commented on libcxx/test/std/containers/views/mdspan/mdspan/properties.pass.cpp:
```diff
@@ -141,12 +141,12 @@ constexpr void test_mdspan_types(const H& handle, const M& map, const A& acc) {
-  assert(!noexcept(m.is_unique()));
-  assert(!noexcept(m.is_exhaustive()));
-  assert(!noexcept(m.is_strided()));
+  LIBCPP_STATIC_ASSERT(!noexcept(MDS::is_always_unique()));
```
@crtrott Why have we added these tests in the first place? And why doesn't the standard require them to be `noexcept`? From what I can tell, these should meet the Lakos rule requirements. Same questions for the other mdspan changes.

StephanTLavavej replied:
@crtrott can answer definitively, but it appeared to me that these tests were semi-mechanically translating the Standardese, and it was a simple oversight that implementations don't have to follow the absence of `noexcept` in the Standard.

For `mdspan::is_always_meow()`, there's a simple proof that the Standard could absolutely mark them as unconditionally `noexcept`. [N4964](https://isocpp.org/files/papers/N4964.pdf) \[mdspan.mdspan.overview\] requires `mdspan::is_always_meow()` to be effects-equivalent-to `{ return mapping_type::is_always_meow(); }`, then the layout mapping requirements in \[mdspan.layout.reqmts\]/22, /24, /26 require `mapping_type::is_always_meow()` to be a constant expression. This could likely be resolved through a simple LWG issue.

For the other `mdspan` operations involved here (construction/`is_meow`/`stride`), we have conditionally strengthened them, and the Standard doesn't usually depict conditional `noexcept` unless it's really important, so that's likely the reason.

philnik777 reviewed (changes requested):
I'd like to have an answer first, since I feel like I'm missing something here that explains the `noexcept` missing from the standard. If that is the case, we may have to fix the test differently, and the MSVC STL would probably also be wrong in making things `noexcept` unconditionally.

StephanTLavavej committed:
LWG-4021 "mdspan::is_always_meow() should be noexcept"

Also, we can portably upgrade `assert` to `static_assert` when checking `MDS::is_always_meow() == M::is_always_meow()`.

StephanTLavavej committed:
Apply clang-format from CI.

StephanTLavavej commented:
@philnik777 I've gotten an LWG issue number and updated the product/test code accordingly. I've also updated the PR title and description.
